### PR TITLE
Allow disabling lenient IntlDateFormatter parsing

### LINKED ISSUE #14519: Use setLenient(false) when parsing localized date(time)
This is a (multiple allowed):

* [x] bug
* [x] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Version: 4.0.
* Platform and Target: All.

### What you did

At [I18n\DateFormatTrait](https://github.com/cakephp/cakephp/blob/master/src/I18n/DateFormatTrait.php) we have methods to parse strings of date and datetime into objects.

### What happened

When passing invalid dates for some locales, the parser silent fail and return a non-predictive date object.

### What you expected to happen

Parsing should fail.

I think it should expect string localized representing valid dates/datetimes.

### Propose

Disable, or allow us to disable, the lenient mode from IntlDateFormatter, [as pointed in PHP documentation](https://www.php.net/manual/en/intldateformatter.setlenient.php).

Code sample:

```php
// If you don't want to setup pt_BR locale, use a string with '25-12-2000' (25 of december)
// It will be parsed as 2002-01-12
Cake\I18n\I18n::setLocale('pt_BR');

/**
 * Day: 04
 * Month: 21   <-- This is invalid
 * Year: 1980
 */
$invalid = '04/21/1980';
$result = Cake\I18n\Date::parseDate($invalid);

// Expected: null|false
// Result: Cake\I18n\Date Object
// (
//     [time] => 1981-09-04 00:00:00.000000+00:00
//     [timezone] => UTC
//     [fixedNowTime] => 2020-04-29T13:50:04.505974+00:00
// )
debug($result);
```

### PR DESCRIPTION
**Pull request by othercorey**
Closes https://github.com/cakephp/cakephp/issues/14519

### CODE REVIEW

**Comment by othercorey** [comment]
Maybe this should be `enableLenientParsing()` and `disableLenientParsing()` and not worry about clearing. No one is going to need that scenario and this reads easier.

File: src/I18n/DateFormatTrait.php [hunk 1]
```diff
@@ -42,6 +42,15 @@ trait DateFormatTrait
...
     protected static $defaultLocale;
 
+    /**
+     * Whether the IntlDateFormatter is set to lenient for parsing.
+     *
+     * Defaults to null which uses IntlDateFormatter default (true).
+     *
+     * @var bool|null
+     */
+    protected static $lentientParsing;
```
Comment (ADmad) [inline src/I18n/DateFormatTrait.php:52]: Why don't you just use true/false instead of null/false?

Comment (ADmad) [reply src/I18n/DateFormatTrait.php:52]: Using two falsey values to indicate different states is not intuitive.

Comment (othercorey) [reply src/I18n/DateFormatTrait.php]: I was trying not to hide the intldateformatter default but that's too complicated for this.

File: src/I18n/DateFormatTrait.php [hunk 2]
```diff
@@ -42,6 +42,15 @@ trait DateFormatTrait
...
     protected static $defaultLocale;
 
+    /**
+     * Whether lenient parsing is enabled for IntlDateFormatter.
+     *
+     * Defaults to true which is the default for IntlDateFormatter.
+     *
+     * @var bool
+     */
+    protected static $lentientParsing = true;
```
Comment (markstory) [inline src/I18n/DateFormatTrait.php:52]:
```suggestion
    protected static $lenientParsing = true;
```

File: src/I18n/DateFormatTrait.php [hunk 3]
```diff
@@ -69,14 +78,46 @@ public static function getDefaultLocale(): ?string
...
     {
         static::$defaultLocale = $locale;
     }
 
+    /**
+     * Gets whether locale format parsing is set to lenient.
+     *
+     * @return bool
+     */
+    public static function lentientParsingEnabled(): bool
```
Comment (markstory) [inline src/I18n/DateFormatTrait.php:96]:
```suggestion
    public static function lenientParsingEnabled(): bool
```

File: src/I18n/DateFormatTrait.php [hunk 4]
```diff
@@ -69,14 +78,46 @@ public static function getDefaultLocale(): ?string
...
+    {
+        return static::$lentientParsing;
```
Comment (markstory) [inline src/I18n/DateFormatTrait.php:98]:
```suggestion
        return static::$lenientParsing;
```

File: src/I18n/DateFormatTrait.php [hunk 5]
```diff
@@ -69,14 +78,46 @@ public static function getDefaultLocale(): ?string
...
+    }
+
+    /**
+     * Enables lentient parsing for locale formats.
```
Comment (markstory) [inline src/I18n/DateFormatTrait.php:102]:
```suggestion
     * Enables lenient parsing for locale formats.
```

File: src/I18n/DateFormatTrait.php [hunk 6]
```diff
@@ -69,14 +78,46 @@ public static function getDefaultLocale(): ?string
...
+     *
+     * @return void
+     */
+    public static function enableLentientParsing(): void
```
Comment (markstory) [inline src/I18n/DateFormatTrait.php:106]:
```suggestion
    public static function enableLenientParsing(): void
```

File: src/I18n/DateFormatTrait.php [hunk 7]
```diff
@@ -69,14 +78,46 @@ public static function getDefaultLocale(): ?string
...
+    {
+        static::$lentientParsing = true;
```
Comment (markstory) [inline src/I18n/DateFormatTrait.php:108]:
```suggestion
        static::$lenientParsing = true;
```

File: src/I18n/DateFormatTrait.php [hunk 8]
```diff
@@ -69,14 +78,46 @@ public static function getDefaultLocale(): ?string
...
+    }
+
+    /**
+     * Enables lentient parsing for locale formats.
```
Comment (markstory) [inline src/I18n/DateFormatTrait.php:112]:
```suggestion
     * Enables lenient parsing for locale formats.
```

File: src/I18n/DateFormatTrait.php [hunk 9]
```diff
@@ -69,14 +78,46 @@ public static function getDefaultLocale(): ?string
...
+     *
+     * @return void
+     */
+    public static function disableLentientParsing(): void
```
Comment (markstory) [inline src/I18n/DateFormatTrait.php:116]:
```suggestion
    public static function disableLenientParsing(): void
```

File: src/I18n/DateFormatTrait.php [hunk 10]
```diff
@@ -69,14 +78,46 @@ public static function getDefaultLocale(): ?string
...
+    {
+        static::$lentientParsing = false;
```
Comment (markstory) [inline src/I18n/DateFormatTrait.php:118]:
```suggestion
        static::$lenientParsing = false;
```

File: src/I18n/DateFormatTrait.php [hunk 11]
```diff
@@ -335,6 +376,8 @@ public static function parseDateTime(string $time, $format = null, $tz = null)
             null,
             $pattern ?? ''
         );
+        $formatter->setLenient(static::$lentientParsing);
```
Comment (markstory) [inline src/I18n/DateFormatTrait.php:379]:
```suggestion
        $formatter->setLenient(static::$lenientParsing);
```

File: tests/TestCase/I18n/DateTest.php [hunk 12]
```diff
@@ -216,6 +216,25 @@ public function testParseDateTime($class)
...
+     * Tests disabling leniency when parsing locale format.
+     *
+     * @dataProvider classNameProvider
+     * @return void
+     */
+    public function testLenientParseDate($class)
+    {
+        $class::setDefaultLocale('pt_BR');
+
+        $class::disableLentientParsing();
```
Comment (markstory) [inline tests/TestCase/I18n/DateTest.php:229]:
```suggestion
        $class::disableLenientParsing();
```

File: tests/TestCase/I18n/DateTest.php [hunk 13]
```diff
@@ -216,6 +216,25 @@ public function testParseDateTime($class)
...
+        $date = $class::parseDate('04/21/2013');
+        $this->assertSame(null, $date);
+
+        $class::enableLentientParsing();
```
Comment (markstory) [inline tests/TestCase/I18n/DateTest.php:233]:
```suggestion
        $class::enableLenientParsing();
```

File: tests/TestCase/I18n/TimeTest.php [hunk 14]
```diff
@@ -916,6 +916,25 @@ public function testParseTime($class)
...
+     * Tests disabling leniency when parsing locale format.
+     *
+     * @dataProvider classNameProvider
+     * @return void
+     */
+    public function testLenientParseDate($class)
+    {
+        $class::setDefaultLocale('pt_BR');
+
+        $class::disableLentientParsing();
```
Comment (markstory) [inline tests/TestCase/I18n/TimeTest.php:929]:
```suggestion
        $class::disableLenientParsing();
```

File: tests/TestCase/I18n/TimeTest.php [hunk 15]
```diff
@@ -916,6 +916,25 @@ public function testParseTime($class)
...
+        $time = $class::parseDate('04/21/2013');
+        $this->assertSame(null, $time);
+
+        $class::enableLentientParsing();
```
Comment (markstory) [inline tests/TestCase/I18n/TimeTest.php:933]:
```suggestion
        $class::enableLenientParsing();
```

**Pull request merged**